crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -1233,59 +1248,105 @@ impl Loader<HistoricalKey> for Db {
+                    }
+
+                    query
+                })
+                .await
+            }
+            .scope_boxed()
+        });
+        // Use a separate query instead of join since this will be a KV query in the future.
+        let native_object_query = self.execute(move |conn| {
+            async {
+                conn.results(move || {
+                    let mut query = f::full_objects_history
+                        .select(StoredFullHistoryObject::as_select())
+                        .into_boxed();
+
+                    for (id, version) in id_versions.iter().cloned() {
+                        query =
+                            query.or_filter(f::object_id.eq(id).and(f::object_version.eq(version)));
+                    }
+
+                    query
+                })
+                .await
+            }
+            .scope_boxed()
+        });
+        let (indexed_objects_result, native_objects_result) =
+            futures::join!(indexed_objects_query, native_object_query);
+        let indexed_objects: Vec<StoredHistoryObject> = indexed_objects_result
+            .map_err(|e| Error::Internal(format!("Failed to fetch objects: {e}")))?;
+        let native_objects: Vec<StoredFullHistoryObject> = native_objects_result
             .map_err(|e| Error::Internal(format!("Failed to fetch objects: {e}")))?;
 
-        let mut id_version_to_stored = BTreeMap::new();
-        for stored in objects {
-            let key = (addr(&stored.object_id)?, stored.object_version as u64);
-            id_version_to_stored.insert(key, stored);
+        let mut id_version_to_indexed = BTreeMap::new();
+        for indexed in indexed_objects {
+            let key = (addr(&indexed.object_id)?, indexed.object_version as u64);
+            id_version_to_indexed.insert(key, indexed);
+        }
+        let mut id_version_to_native = BTreeMap::new();
+        for native in native_objects {
+            let key = (addr(&native.object_id)?, native.object_version as u64);
+            id_version_to_native.insert(key, native);
         }
 
         let mut result = HashMap::new();
         for key in keys {
-            let Some(stored) = id_version_to_stored.get(&(key.id, key.version)) else {
-                continue;
-            };
+            let indexed_opt = id_version_to_indexed.get(&(key.id, key.version));
 
             // Filter by key's checkpoint viewed at here. Doing this in memory because it should be
             // quite rare that this query actually filters something, but encoding it in SQL is
             // complicated.
-            if key.checkpoint_viewed_at < stored.checkpoint_sequence_number as u64 {
-                continue;
-            }
+            let object = if let Some(indexed) = indexed_opt {
+                if key.checkpoint_viewed_at < indexed.checkpoint_sequence_number as u64 {
+                    continue;
+                }
```

Comment:
At first I was confused why we were querying both of these tables here. I think I understand why now, and this part is the key -- we are relying on this query to ensure that the query remains consistent at the tip. How will this in work in our KV store future?